lib/ui/window.dart

```diff
@@ -143,14 +161,240 @@ class Locale {
   /// ```
   ///
   /// The primary language subtag must not be null. The region subtag is
-  /// optional.
-  ///
-  /// The values are _case sensitive_, and should match the case of the relevant
-  /// subtags in the [IANA Language Subtag
-  /// Registry](https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry).
-  /// Typically this means the primary language subtag should be lowercase and
-  /// the region subtag should be uppercase.
-  const Locale(this._languageCode, [ this._countryCode ]) : assert(_languageCode != null);
+  /// optional. The values are _case sensitive_, and must match the values
+  /// listed in
+  /// http://unicode.org/repos/cldr/tags/latest/common/validity/language.xml and
+  /// http://unicode.org/repos/cldr/tags/latest/common/validity/region.xml.
+  ///
+  /// This method only produces standards-compliant instances if valid language
+  /// and country codes are provided. Deprecated subtags will be replaced, but
+  /// incorrectly cased strings are not corrected.
+  const Locale(
+    this._languageCode, [
+    this._countryCode,
+  ]) : assert(_languageCode != null),
+       this._scriptCode = null,
+       this._variants = null,
+       this._extensions = null;
+
+  /// Creates a new Locale object with the specified parts.
+  ///
+  /// This is for internal use only. All fields must already be normalized, must
+  /// already be canonicalized. This method does not modify parameters in any
+  /// way or do any syntax checking.
+  ///
+  /// * language, script and region must be in canonical form.
+  /// * Iterating over variants must provide variants in alphabetical order. We
+  ///   force this by taking a SplayTreeSet as parameter input, which
+  ///   automatically sorts its items. We then convert to a LinkedHashSet for
+  ///   faster access and iteration.
+  /// * The extensions map must contain only valid key/value pairs. "u" and "t"
+  ///   keys must be present, with an empty string as value, if there are any
+  ///   subtags for those singletons.
+  Locale._internal(
```

Comment:
is there any reason this has to be a private constructor? it seems like exposing a way to build a Locale with all the various components is useful. (In particular, isn't that what we're going to want to do when the OS gives us a language, script, and country?)

Reply:
This constructor isn't safe for general use, does not do the necessary checking to maintain API promises. parse() does, on the other hand. I'm only adding parse() in this pull request, with this as a necessary helper to deal with the final members.

Happy to add more as soon as we've decided what other factory methods we want.